plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/WorkspaceConfigSnapshot.java

```diff
@@ -40,38 +44,44 @@ public WorkspaceConfigSnapshot(XIWorkspaceConfig workspace) {
 	}
 
 	/** See {@link WorkspaceConfigSnapshot}. */
-	public WorkspaceConfigSnapshot(URI path, Iterable<? extends IProjectConfigSnapshot> projects) {
-		Map<String, IProjectConfigSnapshot> lookupName2Project = new HashMap<>();
-		Map<URI, IProjectConfigSnapshot> lookupSourceFolderPath2Project = new HashMap<>();
+	public WorkspaceConfigSnapshot(URI path, Iterable<? extends ProjectConfigSnapshot> projects) {
+		Map<String, ProjectConfigSnapshot> lookupName2Project = new HashMap<>();
+		Map<URI, ProjectConfigSnapshot> lookupSourceFolderPath2Project = new HashMap<>();
 		updateLookupMaps(lookupName2Project, lookupSourceFolderPath2Project, projects, Collections.emptyList());
 		this.path = path;
-		this.name2Project = ImmutableMap.copyOf(lookupName2Project);
+		this.name2Project = ImmutableBiMap.copyOf(lookupName2Project);
 		this.sourceFolderPath2Project = ImmutableMap.copyOf(lookupSourceFolderPath2Project);
 	}
 
 	/** See {@link WorkspaceConfigSnapshot}. */
-	protected WorkspaceConfigSnapshot(URI path, ImmutableMap<String, IProjectConfigSnapshot> name2Project,
-			ImmutableMap<URI, IProjectConfigSnapshot> sourceFolderPath2Project) {
+	protected WorkspaceConfigSnapshot(URI path, ImmutableBiMap<String, ProjectConfigSnapshot> name2Project,
+			ImmutableMap<URI, ProjectConfigSnapshot> sourceFolderPath2Project) {
 		this.path = path;
 		this.name2Project = name2Project;
 		this.sourceFolderPath2Project = sourceFolderPath2Project;
 	}
 
-	@Override
+	/**
+	 * Getter for the root path.
+	 */
 	public URI getPath() {
 		return path;
 	}
 
-	@Override
-	public ImmutableCollection<? extends IProjectConfigSnapshot> getProjects() {
+	/**
+	 * Get all the projects known in this snapshot.
+	 */
+	public ImmutableSet<? extends ProjectConfigSnapshot> getProjects() {
```

Comment:
same as above (also for Map vs. BiMap)

Reply:
I think having the same project twice in a workspace will be a source of hard to track-down bugs. Therefore I hardened the constraints.